protocol.md

```diff
@@ -0,0 +1,752 @@
+   call chain. The point of this unique traceid is to build a span tree that
+   covers all RPCs supporting a given client-visible transaction, not just
+   tracing of a single logical request through multiple intermediaries.
+3. generate a unique spanid which is used to refer to this specific step in the
+   network
+4. set the ttl to the maximum allowable time for this request. If a response
+   cannot be sent within this time, downstream nodes will cancel the request.
+5. A1 doesn't need any headers, so defaults are used
+6. a "call req" body is generated with service\_name "service B", application
+   payload args, and checksum
+
+R1 receives "call req" from A1, sends "call req" (0x03) to B1:
+
+1. select message id M2 for this frame
+2. copy traceid from incoming message to traceid of new message
+3. copy spanid from incoming message to parentid of new message
+4. generate unique spanid
+5. copy ttl from incoming message to ttl of new message
+6. copy service name, args, and csum data to new message
+
+B1 receives "call req" from R1, sends "call res" (0x04) to R1:
+
+1. use message id M2 for this frame
+2. send args from application response and compute csum
+
+R1 receives "call res" from B1, send "call res" (0x03) to A1:
+
+1. match incoming message id M2 with existing "call req"
+2. use message id M1 for new message
+3. copy args/csum from incoming message
+
+A1 receives "call res" from R1:
+
+1. match incoming message M1 with existing "call req"
+2. deliver args to application
+
+## Framing
+
+All frames of all types use this structure:
+
+```
+----------------------------------------------
+| 0-7  | size:2 | type:1 | reserved:1 | id:4 |
+|------|-------------------------------------|
+| 8-15 | reserved:8                          |
+|------|-------------------------------------|
+| 16+  | payload - based on type             |
+----------------------------------------------
+```
+
+### size:2
+
+Total length of this frame in bytes, including the framing header and body.
+Note that this limits the total size of a frame to 64KiB, even though some of
+the other fields are also specified with 16 bit sizes. Implementations must
+take care to not exceed the total frame size of 64KiB.
+
+### type:1
+
+Type of the payload.  Valid types are:
+
+`0x01` | init req          | First message on every connection must be init
+`0x02` | init res          | Remote response to init req
+`0x03` | call req          | RPC method request
+`0x04` | call res          | RPC method response
+`0x13` | call req continue | RPC request continuation fragment
+`0x14` | call res continue | RPC response continuation fragment
+`0xc0` | cancel            | Cancel an outstanding call req / forward req (no body)
+`0xc1` | claim             | Claim / cancel a redundant request
+`0xd0` | ping req          | Protocol level ping req (no body)
+`0xd1` | ping res          | Ping res (no body)
+`0xff` | error             | Protocol level error.
```

Comment:
The table doesn't render correctly right now. I think GitHub Markdown requires headers for tables.